aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/AwsHttpServletRequest.java

```diff
@@ -607,38 +607,56 @@ protected List<String> getQueryParamValuesAsList(MultiValuedTreeMap<String, Stri
+        return generateParameterMap(qs, config, false);
+    }
+    
+    protected Map<String, String[]> generateParameterMap(MultiValuedTreeMap<String, String> qs, ContainerConfig config, boolean decodeQueryParams) {
         Map<String, String[]> output;
 
         Map<String, List<String>> formEncodedParams = getFormUrlEncodedParametersMap();
 
         if (qs == null) {
             // Just transform the List<String> values to String[]
-            output = formEncodedParams.entrySet().stream()
+            return formEncodedParams.entrySet().stream()
                 .collect(Collectors.toMap(Map.Entry::getKey, (e) -> e.getValue().toArray(new String[0])));
-        } else {
-            Map<String, List<String>> queryStringParams;
-            if (config.isQueryStringCaseSensitive()) {
-                queryStringParams = qs;
-            } else {
-                // If it's case insensitive, we check the entire map on every parameter
-                queryStringParams = qs.entrySet().stream().parallel().collect(
-                    Collectors.toMap(
-                        Map.Entry::getKey,
-                        e -> getQueryParamValuesAsList(qs, e.getKey(), false)
-                    ));
-            }
-
-            // Merge formEncodedParams and queryStringParams Maps
-            output = Stream.of(formEncodedParams, queryStringParams).flatMap(m -> m.entrySet().stream())
-                .collect(
-                    Collectors.toMap(
-                        Map.Entry::getKey,
-                        e -> e.getValue().toArray(new String[0]),
-                        // If a parameter is in both Maps, we merge the list of values (and ultimately transform to String[])
-                        (formParam, queryParam) -> Stream.of(formParam, queryParam).flatMap(Stream::of).toArray(String[]::new)
-                    ));
+        }
 
+        // decode all keys and values in map
+        final MultiValuedTreeMap<String, String> decodedQs = new MultiValuedTreeMap<String, String>();
+        if (decodeQueryParams) {
+            for (Map.Entry<String, List<String>> entry : qs.entrySet()) {
+                String k = decodeValueIfEncoded(entry.getKey());
+                List<String> v = getQueryParamValuesAsList(qs, entry.getKey(), false).stream()
+                        .map(AwsHttpServletRequest::decodeValueIfEncoded)
+                        .collect(Collectors.toList());
+                // addAll in case map has 2 keys that are identical once decoded
+                decodedQs.addAll(k, v);
+            }
+        } else {
+            decodedQs.putAll(qs);
         }
+        
+        Map<String, List<String>> queryStringParams;
+        if (config.isQueryStringCaseSensitive()) {
+            queryStringParams = decodedQs;
+        } else {
+            // If it's case insensitive, we check the entire map on every parameter
+            queryStringParams = decodedQs.entrySet().stream().collect(
```

Comment:
reintroduced it for now

Reply:
Parallel streams actually perform worse than a sequential stream for small stream data sets due to the work involved in splitting up the data set, farming them out to threads, then recombining the results. Here's an article: https://blogs.oracle.com/javamagazine/post/java-parallel-streams-performance-benchmark.

And because I come from the old days where we had to get every last drop out of available memory and CPU because there was no magical cloud scaling :) So it's an old habit to be searching for these types of efficiency gains. I also reworked some usage in getQueryParameters() where arrays and lists were being created and immediately disposed of.

It's not a big deal in the end, but it's a good one to know. Parallel is not always the silver bullet it seems to be.

Reply:
Thanks for sharing - makes sense! If @valerena agrees, we can address that for 2.1.1, I just didn't want to delay the 2.1.0 release any longer.